### DISCUSSION TITLE
Meshing capabilities?

### DISCUSSION THREAD

**Discussion by gvcallen** [hannobraun/fornjot#2319] [Q&A]
Hi all,

It isn't very clear to me what meshing capabilities Fornjot provides, and I'm wondering if someone could help.

For example, if I have a surface defined by rectangles, circles etc. (or a "shell" as it seems Fornjot calls it, however I am still getting the hang of the docs) am I able to mesh this surface into triangles? Or mesh the entire volume defined by the surface into tetrahedrons? How would I approach discretizing such a surface in Fornjot? Or would I need to do so using an external library?

Thanks in advance,
Gary

**Comment by hannobraun** [comment]
Hey @gvcallen!

Any (3D) shape you can represent in Fornjot, you can turn into a triangle mesh, via the [`Triangulate`](https://docs.rs/fj-core/0.49.0/fj_core/algorithms/triangulate/trait.Triangulate.html) trait. See [`Instance::process_model`](https://github.com/hannobraun/fornjot/blob/1d1a6276ff5f99877d153da0fceb20d8f1cb3c47/crates/fj/src/instance.rs#L63-L87), which has some code that does that.

This is one of the most established and stable parts of Fornjot, as triangulation happens in preparation of rendering (so whenever you run one of the test models from the repository), as well as for exporting to the various supported data format (which is part of the test suite). The API for it isn't great though. Improvements are welcome!

Discretizing the whole volume as tetrahedrons is not supported, but I don't see a reason why that couldn't be implemented on top of the Fornjot libraries.

**Reply by gvcallen** [reply]
Awesome! Thanks for the snappy reply @hannobraun . Will give it a go
